File: src/tetris/TPiece.java
package tetris;

import java.awt.Color;
import java.util.Arrays;

import javalib.impworld.WorldScene;
import javalib.worldimages.Posn;

public class TPiece extends APiece {

	static boolean[][] TPIECE_UP = new boolean[][] {
		{ false, true, false, false }, // *  
		{ true, true, true, false },   //***
		{ false, false, false, false },
		{ false, false, false, false }
	};
	static boolean[][] TPIECE_DOWN = new boolean[][] {
		{ false, false, false, false }, // 
		{ true, true, true, false },    // ***
		{ false, true, false, false },  //  *
		{ false, false, false, false }
	};
	static boolean[][] TPIECE_RIGHT = new boolean[][] {
		{ false, true, false, false }, // * 
		{ false, true, true, false },  // **
		{ false, true, false, false }, // *
		{ false, false, false, false }
	};
	static boolean[][] TPIECE_LEFT = new boolean[][] {
		{ false, true, false, false }, // *
		{ true, true, false, false },  //** 
		{ false, true, false, false }, // *
		{ false, false, false, false }
	};
	

	TPiece(Posn posn) {
		super(posn, new CycleIndex(TPIECE_UP.clone(), TPIECE_LEFT.clone(), TPIECE_RIGHT.clone(), TPIECE_DOWN.clone()), Tetrimino.T);
	}

	
	public boolean hasSpun(Board b) {
		Residue[] top = b.residue.get(this.position.y);
		Residue[] bottom = b.residue.get(this.position.y + 2);
		
		if (this.position.x + 2 > 9 || this.position.x < 0) {
			return false;
		}
		boolean topBlockPresent = !top[this.position.x].equals(Residue.EMPTY) || !top[this.position.x + 2].equals(Residue.EMPTY);
		System.out.println(topBlockPresent);
		boolean bottomBlockPresent = !bottom[this.position.x].equals(Residue.EMPTY) && !bottom[this.position.x + 2].equals(Residue.EMPTY);
		System.out.println(bottomBlockPresent);
		return topBlockPresent && bottomBlockPresent;
	}
	
	public boolean[][] rotInitialState(String s) {
		switch (s) {
		case "up": return TPIECE_UP;
		case "left": return TPIECE_LEFT;
		case "right": return TPIECE_RIGHT;
		default: return TPIECE_DOWN;
		}
	}

}

File: typescript/src/pieces/TPiece.ts
import { Board } from '@/board';
import { APiece } from '@/pieces/APiece';
import { CycleIndex } from '@/pieces/CycleIndex';
import { Residue } from '@/pieces/Residue';
import { Tetrimino } from '@/pieces/Tetrimino';
import { Posn } from 'impworld';

export class TPiece extends APiece {
  static TPIECE_UP: boolean[][] = [
    [false, true, false, false], // *
    [true, true, true, false], //***
    [false, false, false, false],
    [false, false, false, false],
  ];
  static TPIECE_DOWN: boolean[][] = [
    [false, false, false, false], //
    [true, true, true, false], // ***
    [false, true, false, false], //  *
    [false, false, false, false],
  ];
  static TPIECE_RIGHT: boolean[][] = [
    [false, true, false, false], // *
    [false, true, true, false], // **
    [false, true, false, false], // *
    [false, false, false, false],
  ];
  static TPIECE_LEFT: boolean[][] = [
    [false, true, false, false], // *
    [true, true, false, false], //**
    [false, true, false, false], // *
    [false, false, false, false],
  ];

  constructor(posn: Posn) {
    super(
      posn,
      new CycleIndex(
        TPiece.TPIECE_UP.slice(),
        TPiece.TPIECE_LEFT.slice(),
        TPiece.TPIECE_RIGHT.slice(),
        TPiece.TPIECE_DOWN.slice(),
      ),
      Tetrimino.T,
    );
  }
  public hasSpun(b: Board): boolean {
    const top: Residue[] = b.residue[this.position.y];
    const bottom: Residue[] =
      this.position.y + 2 >= 20 ? b.newEmptyRow() : b.residue[this.position.y + 2];
    if (this.position.x + 2 > 9 || this.position.x < 0) {
      return false;
    }
    const topBlockPresent: boolean =
      top[this.position.x] !== Residue.EMPTY || top[this.position.x + 2] !== Residue.EMPTY;
    console.log(topBlockPresent);
    const bottomBlockPresent: boolean =
      bottom[this.position.x] !== Residue.EMPTY && bottom[this.position.x + 2] !== Residue.EMPTY;
    console.log(bottomBlockPresent);
    return topBlockPresent && bottomBlockPresent;
  }

  public rotInitialState(s: string): boolean[][] {
    switch (s) {
      case 'up':
        return TPiece.TPIECE_UP;
      case 'left':
        return TPiece.TPIECE_LEFT;
      case 'right':
        return TPiece.TPIECE_RIGHT;
      default:
        return TPiece.TPIECE_DOWN;
    }
  }
}
